Quaver.Shared/Screens/Gameplay/Rulesets/Keys/Playfield/Hits/Hit.cs

```diff
@@ -0,0 +1,211 @@
+
+        /// <summary>
+        ///     Computes the hit position and the perfect hit position and returns them as a tuple.
+        /// </summary>
+        /// <param name="manager"></param>
+        /// <param name="hitStat"></param>
+        /// <returns>a tuple of (Position, PerfectPosition)</returns>
+        public static (long, long) ComputePositions(HitObjectManagerKeys manager, HitStat hitStat)
+        {
+            long position, perfectPosition;
+
+            var info = hitStat.HitObject;
+            if (hitStat.KeyPressType == KeyPressType.Release
+                || (hitStat.KeyPressType == KeyPressType.None && hitStat.Judgement == Judgement.Okay))
+            {
+                perfectPosition = manager.GetPositionFromTime(info.EndTime);
+
+                if (hitStat.KeyPressType == KeyPressType.None)
+                    position = perfectPosition;
+                else
+                    position = manager.GetPositionFromTime(info.EndTime - hitStat.HitDifference);
+            }
+            else
+            {
+                perfectPosition = manager.GetPositionFromTime(info.StartTime);
+
+                if (hitStat.KeyPressType == KeyPressType.None)
+                    position = perfectPosition;
+                else
+                    position = manager.GetPositionFromTime(info.StartTime - hitStat.HitDifference);
+            }
+
+            return (position, perfectPosition);
+        }
+
+        /// <summary>
+        ///     Computes the first visible position for the hit display. Used for pooling.
+        /// </summary>
+        /// <param name="manager"></param>
+        /// <param name="hitStat"></param>
+        /// <returns></returns>
+        public static long FirstVisiblePosition(HitObjectManagerKeys manager, HitStat hitStat)
+        {
+            var positions = ComputePositions(manager, hitStat);
+            return Math.Min(positions.Item1, positions.Item2);
+        }
+
+        /// <summary>
+        ///     Creates and initializes a Hit.
+        /// </summary>
+        /// <param name="ruleset"></param>
+        /// <param name="hitStat"></param>
+        public Hit(GameplayRulesetKeys ruleset, HitStat hitStat)
+        {
+            Info = hitStat.HitObject;
+            Manager = (HitObjectManagerKeys) ruleset.HitObjectManager;
+            JudgeColors = SkinManager.Skin.Keys[ruleset.Mode].JudgeColors;
+
+            var lane = Info.Lane - 1;
+            var playfield = (GameplayPlayfieldKeys) ruleset.Playfield;
+            ScrollDirection = playfield.ScrollDirections[lane];
+            LineHitPosition = playfield.TimingLinePositionY[lane];
+
+            var laneX = playfield.Stage.Receptors[lane].X;
+            LineToPerfect = new Sprite
+            {
+                Alignment = Alignment.TopLeft,
+                Position = new ScalableVector2(laneX + 0.5f * playfield.LaneSize, 0),
+                Size = new ScalableVector2(1, 0),
```

Comment:
It's typically not a good idea to use 1px size for lines right now, since the entire line can just disappear and cause a flashing effect. Use 2 instead.